### PR TITLE
Example of State Management via ASP.Net and DI

### PR DESCRIPTION
**Pull request by cleemullins**
This sample shows the integration of ASP.Net and the new State Code via a Sample. 

In this sample we have

1. Created the MyBotAccessors as a Singleton via the ASP.Net Core DI infrastructure. 
2. Created the LuisRecognizer as a Singleton via the ASP.Net Core DI infrastructure.
3. Setup the MyBot class, which is Transient from a DI perspective, to take both as constructor arguments. 
4. Updated the Luis Recognizer (and related config files) to align with the new code from the LUIS team. 
5. Made minor changes to the BotFrameworkOptions integration class to expose state constructs.  

The dialog integration is currently broken as the underlying code is in flux. We've moved the Dialogs to a dedicated class, but the routing and integration with Luis is waiting for changes to happen.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/BotStateSet.cs [hunk 1]
```diff
@@ -14,25 +14,27 @@ namespace Microsoft.Bot.Builder
     /// </summary>
     public class BotStateSet : IMiddleware
     {
-        private readonly List<BotState> _botStates = new List<BotState>();
+        //private readonly List<BotState> _botStates = new List<BotState>();
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotStateSet.cs]: Remove

File: libraries/Microsoft.Bot.Builder/BotState.cs [hunk 2]
```diff
@@ -331,4 +331,28 @@ public Task SetAsync(ITurnContext turnContext, T value)
             }
         }
     }
+
+    /// <summary>
+    /// Provides helper methods for getting state objects from the turn context.
+    /// </summary>
+    public static class StateTurnContextExtensions
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotState.cs]: One class per file.

File: libraries/Microsoft.Bot.Builder/BotState.cs [hunk 3]
```diff
@@ -331,4 +331,28 @@ public Task SetAsync(ITurnContext turnContext, T value)
...
+        /// Gets a conversation state property.
+        /// </summary>
+        /// <typeparam name="TState">The type of the state object to get.</typeparam>
+        /// <param name="context">The context object for this turn.</param>
+        /// <param name="accessor">The state accessor object for this property.</param>
+        /// <returns>The state object.</returns>
+        public static async Task<TState> GetStateAsync<TState>(this ITurnContext context, IStatePropertyAccessor<TState> accessor)
+            where TState : class, new()
+        {
+            return await accessor.GetAsync(context).ConfigureAwait(false);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotState.cs]: throw on null state accessor

File: libraries/Microsoft.Bot.Builder/ConversationState.cs [hunk 4]
```diff
@@ -18,5 +18,13 @@ public ConversationState(IStorage storage)
         }
 
         protected override string GetStorageKey(ITurnContext context) => $"conversation/{context.Activity.ChannelId}/{context.Activity.Conversation.Id}";
+
+        ///// <summary>
+        ///// Gets the conversation state property from turn context.
+        ///// </summary>
+        ///// <param name="context">The context object for this turn.</param>
+        ///// <returns>The coversation state object.</returns>
+        //public static TState Get(ITurnContext context) { return context.Services.Get<TState>(PropertyName); }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/ConversationState.cs]: remove

File: libraries/Microsoft.Bot.Builder/Integration/BotFrameworkOptions.cs [hunk 5]
```diff
@@ -43,6 +44,17 @@ public BotFrameworkOptions()
         /// <seealso cref="BotAdapter.Use(IMiddleware)"/>
         public IList<IMiddleware> Middleware { get; } = new List<IMiddleware>();
 
+        /// <summary>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/Integration/BotFrameworkOptions.cs]: Update comment. Currently copy/paste of Middleware comment

File: libraries/Microsoft.Bot.Builder/Integration/BotFrameworkOptions.cs [hunk 6]
```diff
@@ -43,6 +44,17 @@ public BotFrameworkOptions()
...
+        /// Gets a list of the <see cref="IMiddleware"/> to use on each incoming activity.
+        /// </summary>
+        /// <value>The middleware list.</value>
+        /// <seealso cref="BotAdapter.Use(IMiddleware)"/>
+        public IList<BotState> State { get; } = new List<BotState>();
+
+        public UserState UserState => this.State.OfType<UserState>().First();
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/Integration/BotFrameworkOptions.cs]: I don't think we need these anymore. With the accessors now passed into the Bot, the average Bot won't need them.

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotBase.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotBase.cs]: Delete this file.

File: samples-final/8.AspNetCore-LUIS-Bot/Reminder.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace AspNetCore_LUIS_Bot
+{
+    public class Reminder
```
Comment (cleemullins) [inline samples-final/8.AspNetCore-LUIS-Bot/Reminder.cs]: Comments

File: samples-final/8.AspNetCore-LUIS-Bot/ReminderDialog.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
+﻿using System;
```
Comment (cleemullins) [inline samples-final/8.AspNetCore-LUIS-Bot/ReminderDialog.cs]: Copyright headers

File: samples-final/8.AspNetCore-LUIS-Bot/Startup.cs [hunk 10]
```diff
@@ -65,14 +54,71 @@ public void ConfigureServices(IServiceCollection services)
                 //      https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
 
                 // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureTableStorage("AzureTablesConnectionString", "TableName");
-                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                //IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                IStorage dataStore = new MemoryStorage();
 
-                options.Middleware.Add(new ConversationState<Dictionary<string, object>>(dataStore));
-                options.Middleware.Add(new UserState<UserState>(dataStore));
+                // *NEW* CREATE NEW CONVERSATION STATE
```
Comment (vishwacsena) [inline samples-final/8.AspNetCore-LUIS-Bot/Startup.cs]: We are doing user state here but the comment says conversation state?

File: samples-final/8.AspNetCore-LUIS-Bot/Startup.cs [hunk 11]
```diff
@@ -65,14 +54,71 @@ public void ConfigureServices(IServiceCollection services)
...
+                }
+
+                if (string.IsNullOrWhiteSpace(endpointKey))
+                {
+                    throw new InvalidOperationException("The Luis endpoint key ('Luis-EndpointKey') is required to run this sample.");
+                }
+
+                if (string.IsNullOrWhiteSpace(azureRegion))
+                {
+                    throw new InvalidOperationException("The Luis Azure Region ('Luis-AzureRegion') is required to run this sample.");
```
Comment (vishwacsena) [inline samples-final/8.AspNetCore-LUIS-Bot/Startup.cs]: This is cool. For the benefit of a newbie developer that just gets this sample and hits F5, can we have this appended to the exception text - "Please update your appsettings.json."

File: samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+        public MyLuisBot(MyBotAccessors accessors, LuisRecognizer recognizer)
+        {
+            _stateAccessors = accessors ?? throw new ArgumentNullException(nameof(accessors));
+            _luisRecognizer = recognizer ?? throw new ArgumentNullException(nameof(recognizer));
+
+            _dialogSet = new DialogSet();
+
+            var reminderDialog = new ReminderDialog(_stateAccessors);            
+            _dialogSet.Add("AlarmDialog", reminderDialog);
```
Comment (vishwacsena) [inline samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs]: We should have a consistent name here. Typo - alarmDialog and reminderDialog

File: samples-final/8.AspNetCore-LUIS-Bot/Startup.cs (hunk 10 above)
Comment (daveta) [reply samples-final/8.AspNetCore-LUIS-Bot/Startup.cs]: Fixed

File: samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs (hunk 12 above)
Comment (daveta) [reply samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs]: Fixed

File: samples-final/8.AspNetCore-LUIS-Bot/Startup.cs (hunk 11 above)
Comment (daveta) [reply samples-final/8.AspNetCore-LUIS-Bot/Startup.cs]: Fixed, good idea

File: samples-final/8.AspNetCore-LUIS-Bot/appsettings.json [hunk 13]
```diff
@@ -1,7 +1,7 @@
 ﻿{
   "MicrosoftAppId": "",
   "MicrosoftAppPassword": "",
-  "Luis-ModelId": "",
-  "Luis-SubscriptionId": "",
-  "Luis-Url": ""
+  "Luis-ApplicationId": "b78fdf6f-0a58-4510-81dc-bb16cd6902ba",
```
[the remark this answers is not in the document]
Comment (daveta) [reply samples-final/8.AspNetCore-LUIS-Bot/appsettings.json]: Fixed..

**Comment by JonathanFingold** [comment]
See also sample code posted for issue [787](https://github.com/Microsoft/botbuilder-dotnet/issues/787), Proactive Messaging Documentation.

File: samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+using Microsoft.Bot.Schema;
+
+namespace AspNetCore_LUIS_Bot
+{
+    /// <summary>
+    /// MyLuisBot handles the processing of Bot messages (Activity).  This example uses
+    /// the Language Understanding service (LUIS) to assist in analyzing the text received
+    /// and idenfitying intent.  
+    /// 
+    /// A new instance of MyLuisBot is created whenever a new message is received.
```
Comment (daveta) [inline samples-final/8.AspNetCore-LUIS-Bot/MyLuisBot.cs]: >received [](start = 73, length = 8)

We should just state it's transient, which I think is a little different.

**Comment by drub0y** [comment]
So I certainly have some feedback about this, but I think it may bear face to face discussion as it feels like it would just too much detail to pack into a GitHub comment.

#### TL;DR; 
I think adding the `BotFrameworkConfiguration::States` property is a mistake. It doesn't help the framework _do_ anything; it just expanded the surface area with something new and confusing. Just look at the comments on the property right now (already in `master`) and tell me that someone can pick up the framework and figure out what that means. A dev can do what this sample is doing any number of ways that doesn't require adding any additional property to the `BotFrameworkConfiguration`.

**Comment by cleemullins** [comment]
No need for this PR, as all key aspects have been merged in via other PRs.
